### ISSUE TITLE
first level uses only the first design matrix

### ISSUE THREAD

**Issue by satra** [nistats/nistats#350]
> What version of Nistats are you using? 
i was looking at master

> What were you trying to do? 
understand how nistats is modeling run level data

> What did you expect will happen?
that the designs of each run are incorporated

> What actually happened?
only first run design is being used

see: https://github.com/nistats/nistats/blob/master/nistats/first_level_model.py#L552

if the designs are different across runs (see some reasons below), this would not generate appropriate contrasts.

reasons for designs being different across runs:
- number of outliers being included per run
- missing conditions (in a hit/miss experiment, it's possible there are no hits or misses)
- number of noise components (e.g., compcor) being included to explain variance could vary per run

cc: @effigies

**Comment by bthirion** [comment]
Thx for reporting.
If I understand correctly, you want to run a fixed effects analysis on a multi-session dataset, with some missing conditions in some of the sessions ?
I have no good solution for that in the library: what you can do easily is to run independent models on the different sesssions, but then there exists no function that runs the fixed effects analysis on the analysis.
You probably need something like 
https://github.com/hbp-brain-charting/public_analysis_code/blob/master/ibc_public/utils_pipeline.py
l.557-598.
?

**Comment by satra** [comment]
@bthirion - thank you for that pointer. i may use that for other things.

but yes, the main thing is that designs can vary by run within a task of a single subject, and contrasts may need to be combined in specific across runs. we could potentially create a nan contrast when it cannot be computed and then in the fixed effects code use nansum and nanmax

**Comment by AnnaD15** [comment]
I'n not sure but I think my problem is relalated to this one. I have 3 session:
- 1 session: viewing faces and rest
- 2 session: viewing pictures and rest
- 3 session: viewing words and rest
I can compare Faces minus Faces_rest with the code below. However this does not work for pictures and words, probably because faces are my first session and thus the design matrix used. To get the contrast for pictures and words I could add the run number in the following line like this: model.fit(imgs[1] , events[1] , confounds[1] ) by adding the [1] for the picture contrast and [2] for the words. 
This way I got the contrast in each session for each condition, however I'm also interested in comparing faces, pictures and words. When I try computing those 
"zmap = model.compute_contrast('Faces-IAPS', 't',output_type=my_output)"
I get the following error message: "PatsyError: unrecognized token in constraint Faces-IAPS", probably because nistats only makes use of the first design matrix.
What is the correct way to compare those three sessions in Nistats?

Thanks a lot for any recommendations!

```
fig, axes = plt.subplots(nrows=12, ncols=4, figsize=(16, 27))
model_and_args = zip(models, models_run_imgs, models_events, models_confounds)
for midx, (model, imgs, events, confounds) in enumerate(model_and_args):
    # fit the GLM
    model.fit(imgs, events, confounds)
    # compute the contrast of interest
    zmap = model.compute_contrast('Faces-Faces_rest', 't',output_type=my_output)

    plotting.plot_glass_brain(zmap.get(z_score), colorbar=True,
                              title=('sub-' + model.subject_label),
                              axes=axes[int(midx / 4), int(midx % 4)],
                              plot_abs=False, display_mode='x', symmetric_cbar=True)
fig.suptitle('subjects memorizing faces network (unc p<0.001)')
plotting.show()
```

**Comment by bthirion** [comment]
Yes, this seems to be related. Actually, Nistats currently does not support what you want to do.
The only  solution I can propose atm is to
* fit 3 different models on the three sessions
* collect the summary statistics for each session
* create the contrasts manually, by adapting the snapshot of code I sent to @satra 
Does it make sense ?

**Comment by annaxyzx** [comment]
Thanks for the quick answer! I'm not sure if this helps me (but also if I understand the code correctly). I thought this aproach would only be useful, when I want to get the mean effects across session? But as every condition is in only one of those sessions, could I 
- fit the 3 different models on the three sessions 
- and collect the summary statistics for each session as you suggested
but then finally contrast the 3 models somehow similiar as in https://github.com/nistats/nistats/issues/162 ? However this did not work either as the model resulted in df=0.. 
What do you think? Or did I missunderstand something?
Thanks a lot!

**Comment by annaxyzx** [comment]
or probably I could for this purpose use the contrast image (effect size images?) of the first level contrasts for each session and compare them similiar to this example:
https://nistats.github.io/auto_examples/03_second_level_models/plot_second_level_two_sample_test.html#sphx-glr-auto-examples-03-second-level-models-plot-second-level-two-sample-test-py ?

**Comment by bthirion** [comment]
you should use the following rules (computation of a fixed effects contrast):
assume that you want to compute x1 - x2, where x1 is an effect observed in session 1 and x2 an effect observed in session 2
For x1, you've got an estimate b1, and a variance v1
For x2, you've got an estimate b2, and a variance v2

then the estimate of x1 - x2 is b = b1 - b2 and the corresponding variance is v = v1 + v2
then you can compute a t statistic t = b  / sqrt(v), with d degrees of freedom
d = d1 + d2, where d1 are those of v1, d2 those of v2 (in practice this should be a large number).

You don't want to fit a random effects model, since you only have two observations, hence cannot estimate properly the variance.

HTH,

**Comment by annaxyzx** [comment]
thanks a lot for the detailed answer!!

**Comment by kchawla-pi** [comment]
I think this will be a good idea for a full example.

**Comment by jeromedockes** [comment]
Related to these issues, since it is not supported maybe we should remove accepting a list of contrasts in `compute_contrast`?  

 `nistats.first_level_model.FirstLevelModel.compute_contrast` accepts as the
   contrast definition a list specifying a different contrast for each run. This
   is covered by unit tests, but I could not see it used in any example, in what
   practical situations does the need for this arise? If it is very rare, maybe
   the columns of the different design matrices can be adjusted so that the same
   contrast serves for all runs?

   I find it a bit confusing because the GLM reporter also accepts a list of
   contrasts, but in the reporter case they correspond to truly different
   contrasts, that are computed separately, using all runs for each contrasts.
   ie with the GLM reporter the number of computed maps is the number of
   provided contrasts, whereas with the first level model the number of computed
   maps is 1 and the number of contrasts must be broadcastable to the number of
   runs.

**Comment by bthirion** [comment]
Sometimes, the design matrix has different sumber of columns per run 
(because some conditions are not present).
I agree it's not the most straightforward case, and it is indeed not 
illustrated in our examples, but it might occur.

What is the problem with keeping it  ?

Best,

**Comment by jeromedockes** [comment]
there is no problem but from what I understand at the moment in this case it
will fail if the design matrices actually have different columns and a list of
strings is passed, it would need to be changed to something like #416 (wip)

moreover, then the documentation of the GLM reporter should probably be
improved:

https://github.com/nistats/nistats/blob/3ffd1427c86d12745957be2bd4cdb9cd14ac3c4e/nistats/reporting/glm_reporter.py#L96

contrasts in GLM reporter are passed one by one in separate calls to
`compute_contrast`. providing several contrasts to `make_glm_report` or to
`compute_contrast` has different semantics.

**Comment by bthirion** [comment]
OK, and then I agree, it would make more sense that multiple contrasts are all related to the same GLM, and that by default design matrices are assumed identical across sessions.
Note that we now have an explicit fixed_effects routine that can be used to issue statistics in the case where the contrasts vectors would differ across sessions.
